kork/kork-artifacts/src/main/java/com/netflix/spinnaker/kork/artifacts/artifactstore/entities/MapSerializerHook.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2025 Apple Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.kork.artifacts.artifactstore.entities;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.BeanProperty;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.ser.std.MapSerializer;
+import com.netflix.spinnaker.kork.artifacts.artifactstore.ArtifactStore;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * MapSerializerHook will hook into the map serializers to match a map to a handler.
+ *
+ * <p>This class will not serialize directly, and all it will do is prep the map to be serialized.
+ * The reasoning for this is there are a lot of ways to plug into jackson, and especially object
+ * mappers. To retain the customized behavior, we can only affect the object itself before it needs
+ * to be serialized.
+ */
+public class MapSerializerHook extends MapSerializer {
+  /**
+   * This is the serializer that we maintain and that was passed during the modifyMapSerializer. It
+   * is important to note that MapSerializers are handled VERY differently than StdSerializer or
+   * JsonSerializer. If we were to instead use the inappropriate serializer as a wrapper, the
+   * serializing workflow will not bootstrap the map serializer to properly serialize the incoming
+   * object. A good example of this is when a MapSerializer is passed in, the key serializer may be
+   * null. The serializer workflow will pass in the key serializer by using the withResolved method.
+   */
+  private final MapSerializer defaultSerializer;
+
+  private final ArtifactStore storage;
+  /** Handlers to handle specific keys in the JSON. */
+  private final List<ArtifactHandler> handlers;
+
+  public MapSerializerHook(
+      ArtifactStore storage, List<ArtifactHandler> handlers, MapSerializer defaultSerializer) {
```

Comment:
Check that handlers isn't null?  Otherwise visit could end up throwing a NullPointerException.